model/generate.go

```diff
@@ -275,25 +270,160 @@ func (g *GeneratedProject) saveNewBuildsAndTasks(ctx context.Context, v *Version
 	}
 
 	activatedTasksInExistingBuilds, err := addNewTasks(ctx, activationInfo, v, p, projectRef, newTVPairsForExistingVariants,
-		existingBuilds, syncAtEndOpts, g.TaskID)
+		existingBuilds, syncAtEndOpts, g.Task.Id)
 	if err != nil {
 		return errors.Wrap(err, "adding new tasks")
 	}
 
 	activatedTasksInNewBuilds, err := addNewBuilds(ctx, activationInfo, v, p, newTVPairsForNewVariants,
-		existingBuilds, syncAtEndOpts, projectRef, g.TaskID)
+		existingBuilds, syncAtEndOpts, projectRef, g.Task.Id)
 	if err != nil {
 		return errors.Wrap(err, "adding new builds")
 	}
 
 	// only want to add dependencies to activated tasks
-	if err = addDependencies(t, append(activatedTasksInExistingBuilds, activatedTasksInNewBuilds...)); err != nil {
+	if err = g.addDependencies(append(activatedTasksInExistingBuilds, activatedTasksInNewBuilds...)); err != nil {
 		return errors.Wrap(err, "adding dependencies")
 	}
 
 	return nil
 }
 
+func (g *GeneratedProject) SimulateNewDependencyGraph(v *Version, p *Project, projectRef *ProjectRef) error {
+	existingTasksGraph, err := task.VersionDependencyGraph(g.Task.Version, false)
+	if err != nil {
+		return errors.Wrapf(err, "creating dependency graph for version '%s'", g.Task.Version)
+	}
+
+	simulatedGraph, err := g.addNewTasksToGraph(existingTasksGraph, v, p, projectRef)
+	if err != nil {
+		return errors.Wrap(err, "simulating new tasks")
+	}
+
+	if cycles := simulatedGraph.Cycles(); len(cycles) > 0 {
```

Comment:
Should this function name be something around finding cycles since that's the case where we error (rather than just graph construction)?

Reply:
I renamed the function and changed the signature to return a list of cycles separate from the error. 
One benefit of separating the cycles from the error is that the error could be transient, while the cycles will not be.